corehq/apps/api/es.py

```diff
@@ -701,35 +513,55 @@ def consume_params(self, raw_params):
 ]
 
 
-def es_search(request, domain, reserved_query_params=None):
-    return es_search_by_params(request.GET, domain, reserved_query_params)
+def _validate_and_get_es_filter(search_param):
+    #
+    try:
+        _filter = search_param.pop('filter')
+        # custom use case by 'enveritas' project for Form API
+        date_range = _filter['range']['inserted_at']
+        return {
+            'range': {'inserted_at': date_range}
+        }
+    except KeyError:
+        pass
+    try:
+        # Data export tool passes below structure. validate that it is so
+        _range = None
+        try:
+            _range = _filter['or'][0]['and'][0]['range']['server_modified_on']
+        except KeyError:
+            try:
+                _range = _filter['or'][0]['and'][1]['range']['server_modified_on']
+            except KeyError:
+                pass
+
+        if _range:
+            return XFormServerModifiedParams().consume_params({'server_modified_on': _range})
+        else:
+            raise Http400
+    except (KeyError, AssertionError):
+        raise Http400
 
 
-def es_search_by_params(search_params, domain, reserved_query_params=None):
-    payload = {
-        "filter": {
-            "and": [
-                {"term": {"domain.exact": domain}}
-            ]
-        },
-    }
+def es_query_from_get_params(search_params, domain, reserved_query_params=None, doc_type='form'):
+    # doc_type can be form or case
+    assert doc_type in ['form', 'case']
+    es = FormES() if doc_type == 'form' else CaseES()
 
-    # ?_search=<json> for providing raw ES query, which is nonetheless restricted here
-    # NOTE: The fields actually analyzed into ES indices differ somewhat from the raw
-    # XML / JSON.
-    if '_search' in search_params:
-        additions = json.loads(search_params['_search'])
-
-        if 'filter' in additions:
-            payload['filter']['and'].append(additions['filter'])
+    query = es.remove_default_filters().domain(domain)
 
-        if 'query' in additions:
-            payload['query'] = additions['query']
+    if doc_type == 'form':
+        if 'include_archived' in search_params:
+            query = query.filter(
+                filters.OR(filters.term('doc_type', 'xforminstance'), filters.term('doc_type', 'xformarchived')))
+        else:
+            query = query.filter(filters.term('doc_type', 'xforminstance'))
 
-    # ?q=<lucene>
-    if 'q' in search_params:
-        payload['query'] = payload.get('query', {})
-        payload['query']['query_string'] = {'query': search_params['q']}  # A bit indirect?
+    if '_search' in search_params:
+        # This is undocumented usecase by Data export tool and one custom project
+        #   Validate that the passed in param is one of these two expected
+        _filter = _validate_and_get_es_filter(json.loads(search_params['_search']))
+        query = query.filter(_filter)
```

Comment:
this should this check if the returned value is `None`

Reply:
Sure

Reply:
Actually, that method should either return a valid filter or raise 400. Ignoring a None would mean ignoring the `_search` param of user input.